homeassistant/components/starline/config_flow.py

```diff
@@ -0,0 +1,276 @@
+        """Initialize flow."""
+        self._app_id: Optional[str] = None
+        self._app_secret: Optional[str] = None
+        self._username: Optional[str] = None
+        self._password: Optional[str] = None
+        self._mfa_code: Optional[str] = None
+
+        self._app_code = None
+        self._app_token = None
+        self._user_slid = None
+        self._user_id = None
+        self._slnet_token = None
+        self._slnet_token_expires = None
+        self._captcha_image = None
+        self._captcha_sid = None
+        self._captcha_code = None
+        self._phone_number = None
+
+        self._auth = StarlineAuth()
+
+    @staticmethod
+    @callback
+    def async_get_options_flow(config_entry):
+        """Get the options flow for this handler."""
+        return StarlineOptionsFlowHandler(config_entry)
+
+    async def async_step_user(self, user_input=None):
+        """Handle a flow initialized by the user."""
+        return await self.async_step_init(user_input)
+
+    async def async_step_init(self, user_input=None):
+        """Handle a flow initialized by the user."""
+        return await self.async_step_auth_app(user_input)
+
+    async def async_step_auth_app(self, user_input=None, error=None):
+        """Authenticate application step."""
+        if user_input is not None:
+            self._app_id = user_input[CONF_APP_ID]
+            self._app_secret = user_input[CONF_APP_SECRET]
+            return await self._async_authenticate_app(error)
+        return self._async_form_auth_app(error)
+
+    async def async_step_auth_user(self, user_input=None, error=None):
+        """Authenticate user step."""
+        if user_input is not None:
+            self._username = user_input[CONF_USERNAME]
+            self._password = user_input[CONF_PASSWORD]
+            return await self._async_authenticate_user(error)
+        return self._async_form_auth_user(error)
+
+    async def async_step_auth_mfa(self, user_input=None, error=None):
+        """Authenticate mfa step."""
+        if user_input is not None:
+            self._mfa_code = user_input[CONF_MFA_CODE]
+            return await self._async_authenticate_user(error)
+        return self._async_form_auth_mfa(error)
+
+    async def async_step_auth_captcha(self, user_input=None, error=None):
+        """Captcha verification step."""
+        if user_input is not None:
+            self._captcha_code = user_input[CONF_CAPTCHA_CODE]
+            return await self._async_authenticate_user(error)
+        return self._async_form_auth_captcha(error)
+
+    def _async_form_auth_app(self, error=None):
+        """Authenticate application form."""
+        errors = {}
+        if error is not None:
+            errors["base"] = error
+
+        return self.async_show_form(
+            step_id="auth_app",
+            data_schema=vol.Schema(
+                {
+                    vol.Required(
+                        CONF_APP_ID, default=self._app_id or vol.UNDEFINED
```

Comment:
We should not ask app ID and secret via config flows. Those should be moved to configuration.yaml.

For an example, see withings: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/withings/__init__.py#L51-L57

Reply:
This is a bad user experience. AppID and secret should not be saved anywhere. They are used only once to receive an API token.

Reply:
@balloob is it really required?
Integration can be configured without `configuration.yaml` and saves only API tokens. I think, it's very good UX.

Reply:
The goal is that things like app ID and secret can be provided via a central Home Assistant service. That way the user won't need to know about it at all and can just log in from the get go. But to get there, we should not have the app ID / secret in the config flow to begin with. The one thing we want to get out of our documentation is the "Now go to the developer console and make a new app".

By having it in configuration.yaml, someone could install Home Assistant and non-tech users can set it up.

Reply:
@balloob unfortunately, no.
One pair of AppID-Secret you can use only with one starline account.
And we need ask AppID only with account credentials.
It's documented on [starline developer page](https://developer.starline.ru/#api-_).
Translated part:
_You need AppID and AppSecret to access the API. You can get AppID and AppSecret in your account on my.starline.ru. These AppID and AppSecret will allow you to control **exclusively your car**._

Reply:
Okay. In that case it's fine .